File: Archive#00-001-100/023-Merge k Sorted Lists/leetcode023.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Dummy(tuple):
    def __lt__(self, other):
        return self[0] < other[0]

import heapq

class Solution:
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        minHeap, head = [Dummy((node.val, node)) for node in lists if node], ListNode(0)
        heapq.heapify(minHeap)
        p = head
        while minHeap:
            _, node = heapq.heappop(minHeap)
            if node.next: heapq.heappush(minHeap, Dummy(((node.next.val, node.next))))
            p.next, node.next = node, None
            p = p.next
        return head.next

File: Archive#00-001-100/023-Merge k Sorted Lists/leetcode023.rb
# Definition for singly-linked list.
# class ListNode
#   attr_accessor :val, :next
#   def initialize(val)
#     @val = val
#     @next = nil
#   end
# end

# @param {ListNode[]} lists
# @return {ListNode}
def merge_k_lists( lists )
  head = ListNode.new(0)
  p = head
  data = []
  lists.each { |e|
    while (e != nil)
      data << e.val
      e = e.next
    end
  }
  data.sort.each { |e|
    p.next = ListNode.new(e)
    p = p.next
  }
  head.next
end